*: add, clean up e2e tests

gyuho opened the pull request.

gyuho commented:
1. added test: `auth enable`, `defrag`, `endpoint status`, `member list`
2. cleaned up by:
 - separate each command to each go file
 - separate `curl` command tests in v2 into a separate file

TODO: `alarm`, `auth disable (not implemented)`, `snapshot`, `user`, `role`, `member (other commands)`

/cc @heyitsanthony @xiang90 Thanks.

heyitsanthony commented on e2e/ctl_v2_curl_test.go:
```diff
@@ -0,0 +1,113 @@
+// Copyright 2016 CoreOS, Inc.
```
Should this file be prefixed with `ctl`? It's not using etcdctl.

heyitsanthony commented on e2e/ctl_v2_curl_test.go:
```diff
@@ -0,0 +1,113 @@
+	"github.com/coreos/etcd/pkg/testutil"
+)
+
+func TestBasicOpsNoTLS(t *testing.T)        { testBasicOpsPutGet(t, &configNoTLS) }
```
Should these be `TestV2Curl...`?

heyitsanthony commented on e2e/ctl_v3_defrag_test.go:
```diff
@@ -0,0 +1,44 @@
+	var kvs = []kv{{"key", "val1"}, {"key", "val2"}, {"key", "val3"}}
+	for i := range kvs {
+		if err := ctlV3Put(cx, kvs[i].key, kvs[i].val, ""); err != nil {
+			cx.t.Fatalf("compactTest #%d: ctlV3Put error (%v)", i, err)
```
`cx.t.Fatal(err)`

heyitsanthony commented on e2e/ctl_v3_defrag_test.go:
```diff
@@ -0,0 +1,44 @@
+			cx.t.Fatalf("compactTest #%d: ctlV3Put error (%v)", i, err)
+		}
+	}
+
```
should there be a compact?

heyitsanthony commented on e2e/ctl_v3_defrag_test.go:
```diff
@@ -0,0 +1,44 @@
+	cmdArgs := append(cx.PrefixArgs(), "defrag")
+	lines := make([]string, cx.epc.cfg.clusterSize)
+	for i, ep := range cx.epc.endpoints() {
+		lines[i] = fmt.Sprintf("Finished defragmenting etcd member[%s]", ep)
```
just `Finished defragmenting etcd member`? Defrag shouldn't be serialized like it is now-- output ordering could change.

gyuho replied:
You are right. will fix.

heyitsanthony commented on e2e/ctl_v3_endpoint_test.go:
```diff
@@ -0,0 +1,46 @@
+import "testing"
+
+func TestCtlV3EndpointHealthQuorum(t *testing.T) { testCtl(t, endpointHealthTest, withQuorum()) }
+func TestCtlV3EndpointStatusQuorum(t *testing.T) { testCtl(t, endpointStatusTest, withQuorum()) }
```
`TestCtlV3EndpointStatus` -- there's no quorum in this command

gyuho replied:
Oh will fix. thanks.

heyitsanthony commented on e2e/etcd_test.go:
```diff
@@ -464,3 +370,18 @@ func (epc *etcdProcessCluster) proxies() []*etcdProcess {
+	eps := make([]string, epc.cfg.clusterSize)
+	for i, pc := range epc.backends() {
+		ep := pc.cfg.acurl
+		if strings.HasPrefix(ep, "http://") {
```
I don't think the `HasPrefix` checks are necessary.

gyuho commented:
@heyitsanthony PTAL.

1. renamed the `curl` test file with `v2_curl_test`
2. change `curl` test names with `TestV2Curl`
3. added compaction operation before defrag command
4. changed `defrag` command test to just check slice of strings without endpoints in it because it's not in order
5. delete `HasPrefix` parts and just use `url.Parse.Host` to get the endpoints without scheme

Thanks.

heyitsanthony commented on e2e/ctl_v3_test.go:
```diff
@@ -119,12 +72,12 @@ func withInteractive() ctlOption {
 }
 
-func setupCtlV3Test(t *testing.T, cfg etcdProcessClusterConfig, quorum bool) *etcdProcessCluster {
+func setupCtlV3Test(t *testing.T, cfg *etcdProcessClusterConfig, quorum bool) *etcdProcessCluster {
```
why the change to `*etcdProcessClusterConfig`?

gyuho commented on e2e/ctl_v3_defrag_test.go:
```diff
@@ -0,0 +1,45 @@
+
+func ctlV3Defrag(cx ctlCtx) error {
+	cmdArgs := append(cx.PrefixArgs(), "defrag")
+	lines := make([]string, cx.epc.cfg.clusterSize)
```
@heyitsanthony I changed it to pass pointer because otherwise this returns 3 even when it's not running with quorum.

heyitsanthony replied:
So the quorum flag isn't being respected? I don't think that is a good way to fix the problem.

gyuho replied:
It is respected, but we just pass around the preconfigured config object and to not manipulate that shared config, we pass them by value in setupCtlV3Test function. I will try to clean up arg types so that we don't do this hacky stuff. Thanks!

heyitsanthony replied:
I think the fix is to move the code in `setupCtlV3Test` into `testCtl` so ret.cfg is set when the quorum flag is checked.

gyuho replied:
Ok good idea. Let me try that. Thanks!

gyuho commented:
@heyitsanthony Just removed `setupCtlV3Test`. PTAL. Thanks!

heyitsanthony commented:
lgtm

gyuho commented:
thanks. merging in greens.
